designs/initial.md

```diff
@@ -0,0 +1,60 @@
+## `@rest` directive
+
+### Arguments
+
+- route: path to rest endpoint. This could be a path or a full url. If a path, add to the endpoint given on link creation or from the context
+- params: a map of variables to url params
+- method: the HTTP method to send the request via (i.e GET, PUT, POST)
+- type: The GraphQL type this will return
+
+### Notes
+
+It's important that the rest directive could be used at any depth in a query, but once it is used, nothing nested in it can be graphql data, it has to be from the rest link or other resource (like a @client directive)
+
+## `@export` directive
+
+### Arguments
+- as: the string name to create this as a variable to be used down the selection set
+
+### Notes
+
+These are the same semantics that will be supported on the server, but when used in a rest link you can use the exported variables for futher calls (i.e. waterfall requests from nested fields)
+
+```js
+const QUERY = gql`
+  query RestData($email: String!) {
+    users @rest(route: '/users/email/:email', params: { email: $email }, method: 'GET', type: 'User') {
+      id @export(as: "id")
+      firstName
+      lastName
+      friends @rest(route: '/friends/:id', params: { id: $id }, type: '[User]') {
+        firstName
+        lastName
+      }
+    }
+  }
+`;
+```
+
+
+## `createRestLink`
+
+### Arguments
+
+- fetch: an optional implementation of `fetch` (see the http-link for api / warnings). Will use global if found
+- fieldNameNormalizer: a function that takes the response field name and turns into a GraphQL compliant name,for instance "MyFieldName:IsGreat" => myFieldNameIsGreat
```

Comment:
- Should we have a fieldValueNormalizer? -- it would allow the user to rewrite a particular response before it gets interpreted.
- Related: should we have a hook that permits the REST link to receive XML or msgpack, and then have the user transform that to JSON before it gets processed as GraphQL?

Reply:
> Should we have a fieldValueNormalizer? -- it would allow the user to rewrite a particular response before it gets interpreted.

hmm this can be done in an afterware style link as needed, how much would this be used for your usecase?

Reply:
> should we have a hook that permits the REST link to receive XML or msgpack, and then have the user transform that to JSON before it gets processed as GraphQL?

so a `parse` handler? Sure that doesn't seem to bad to add in for advanced usage

Reply:
I don't need the `fieldValueNormalizer` today, but I'm paranoid, and that seems like a good escape hatch. It's also symmetric with the `fieldNameNormalizer' and complementary to the `parse` handler. -- I'm also not an expert in the full breadth of APIs I'm looking to adapt this too.

One of the use cases I had for `fieldValueNormalizer` could instead be modeled as a Union of several representations (at the GraphQL) and then normalized at the client layer. (Think Date-type that is sometimes a number, sometimes a string, how do you normalize it to a Date-class-instance?). But, it'd be nicer if the client code could always expect a clean type, and the data layer handles the variants transparently. Thoughts?

I don't personally need the `parse` handler today, but JSON-only seemed a weakness when we're talking to arbitrary REST endpoints, so I totally recommend it as a feature for `apollo-link-rest`.

Reply:
thinking about the `parse` portion and taking from apollo-link-http, passing a custom `fetch` makes this super easy. You can essentially wrap `fetch` and handle funky parsing on the flly

Reply:
doing `fieldValueNormalizer` could be acheived using something akin to cache-resolvers like Apollo Client by using graphql-anywhere, or it could be a single function but it may get out of hand with if/thens?